stan/math/prim/mat/prob/neg_binomial_2_log_glm_lpmf.hpp

```diff
@@ -95,107 +124,94 @@ neg_binomial_2_log_glm_lpmf(const T_y& y, const T_x& x, const T_alpha& alpha,
   if (!include_summand<propto, T_x, T_alpha, T_beta, T_precision>::value)
     return 0.0;
 
-  Array<T_partials_return, Dynamic, 1> y_arr(N, 1);
-  Array<T_partials_return, Dynamic, 1> phi_arr(N, 1);
-  {
-    scalar_seq_view<T_y> y_vec(y);
-    scalar_seq_view<T_precision> phi_vec(phi);
-    for (size_t n = 0; n < N; ++n) {
-      y_arr[n] = y_vec[n];
-      phi_arr[n] = value_of(phi_vec[n]);
-    }
-  }
+  const auto& x_val = value_of_rec(x);
+  const auto& y_val = value_of_rec(y);
+  const auto& beta_val = value_of_rec(beta);
+  const auto& alpha_val = value_of_rec(alpha);
+  const auto& phi_val = value_of_rec(phi);
 
-  Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
-  {
-    scalar_seq_view<T_beta> beta_vec(beta);
-    for (size_t m = 0; m < M; ++m) {
-      beta_dbl[m] = value_of(beta_vec[m]);
-    }
-  }
+  const auto& y_val_vec = as_column_vector_or_scalar(y_val);
+  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto& alpha_val_vec = as_column_vector_or_scalar(alpha_val);
+  const auto& phi_val_vec = as_column_vector_or_scalar(phi_val);
 
-  Array<T_partials_return, Dynamic, 1> theta_dbl
-      = (value_of(x) * beta_dbl).array();
-  scalar_seq_view<T_alpha> alpha_vec(alpha);
+  const auto& y_arr = as_array_or_scalar(y_val_vec);
+  const auto& phi_arr = as_array_or_scalar(phi_val_vec);
+
+  Array<T_partials_return, Dynamic, 1> theta = value_of(x) * beta_val_vec;
+  theta += as_array_or_scalar(alpha_val_vec);
   for (size_t n = 0; n < N; ++n)
-    theta_dbl[n] += value_of(alpha_vec[n]);
-  Array<T_partials_return, Dynamic, 1> log_phi = phi_arr.log();
-  Array<T_partials_return, Dynamic, 1> logsumexp_eta_logphi
-      = theta_dbl.binaryExpr(log_phi, [](const T_partials_return& xx,
-                                         const T_partials_return& yy) {
-          return log_sum_exp(xx, yy);
-        });
-  Array<T_partials_return, Dynamic, 1> y_plus_phi = y_arr + phi_arr;
+    check_finite(function, "Matrix of independent variables", theta[n]);
+  T_precision_val log_phi = log(phi_arr);
+  Array<T_partials_return, Dynamic, 1> logsumexp_theta_logphi
+      = (theta > log_phi)
+            .select(theta + Eigen::log1p(Eigen::exp(log_phi - theta)),
```

Comment:
Will these `Eigen::log1p` and `Eigen::exp` functions delegate back to our base versions for `var` and `fvar`?  If not, they won't work with nested autodiff.

Reply:
When the autodiff type is `fvar<var>`, then the scalar types will be `var`.  We don't actually use that anywhere in our interfaces, but we have experimental uses and hope to expose it soon.  But we can back-fix this stuff easily if it doesn't work.